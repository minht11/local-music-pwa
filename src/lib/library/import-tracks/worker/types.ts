export interface TrackImportCount {
	/** Count of many tracks were newly added */
	newlyImported: number
	/** Count of many existing tracks were updated */
	existingUpdated: number
	/** Count of many tracks were removed */
	removed: number
	/** Index of currently scanned track */
	current: number
	/** Total count of tracks */
	total: number
}

export interface TrackImportMessage {
	finished: boolean
	count: TrackImportCount
}

export type TrackImportOptions =
	| {
			action: 'directory-add' | 'directory-rescan'
			dirId: number
			dirHandle: FileSystemDirectoryHandle
	  }
	| {
			action: 'file-handles-add'
			files: FileSystemFileHandle[]
	  }
	| {
			action: 'files-add'
			files: File[]
	  }
