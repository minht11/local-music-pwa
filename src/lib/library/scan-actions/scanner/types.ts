import type { FileEntity } from '$lib/helpers/file-system'

export interface TracksScanResult {
	/** Count of many tracks were newly added */
	newlyImported: number
	/** Index of currently scanned track */
	current: number
	/** Total count of tracks */
	total: number
}

export interface TracksScanMessage {
	finished: boolean
	count: TracksScanResult
}

/** @public */
export type TracksScanOptions =
	| {
			action: 'directory-add' | 'directory-rescan'
			dirId: number
			dirHandle: FileSystemDirectoryHandle
	  }
	// Used in the browsers which do not support `showDirectoryPicker`
	// or with tracks from previous application version
	| {
			action: 'legacy-files-migrate-from-v1'
			files: { uuid: string; file: FileEntity }[]
	  }
	| {
			action: 'legacy-files-add'
			files: FileEntity[]
	  }
