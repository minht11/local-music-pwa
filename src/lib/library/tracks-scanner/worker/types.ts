import type { FileEntity } from '$lib/helpers/file-system'

export interface TracksScanResult {
	/** Count of many tracks were newly added */
	newlyImported: number
	/** Index of currently scanned track */
	current: number
	/** Total count of tracks */
	total: number
}

export interface TracksScanMessage {
	finished: boolean
	count: TracksScanResult
}

export type TracksScanOptions =
	| {
			action: 'directory-add' | 'directory-rescan'
			dirId: number
			dirHandle: FileSystemDirectoryHandle
	  }
	// Used in the browsers which do not support `showDirectoryPicker`
	// or with tracks from previous application version
	| {
			action: 'legacy-files-migrate-from-prev-app-version'
			files: FileEntity[]
	  }
	| {
			action: 'legacy-files-add'
			files: FileEntity[]
	  }
