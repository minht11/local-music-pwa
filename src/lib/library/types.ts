import type { FileEntity } from '$lib/helpers/file-system.ts'

export type LibraryStoreName = 'tracks' | 'albums' | 'artists' | 'playlists'

/**
 * Used in browsers where `showDirectoryPicker` is not supported.
 * `file` field is gonna be `File` in those browsers,
 * or if user has tracks from previous application version
 * where directories were not used `FileSystemHandle`.
 */
export const LEGACY_NO_NATIVE_DIRECTORY = -1

/** Special type of playlist which user cannot modify */
export const FAVORITE_PLAYLIST_ID = -1
export const FAVORITE_PLAYLIST_UUID = 'favorites'

/**
 * Used to represent unknown Artist/Album and other values inside database
 * Using ~ so when sorting items are always at the end
 */
export const UNKNOWN_ITEM = '~\0unknown'

export type UnknownItem = typeof UNKNOWN_ITEM

export type StringOrUnknownItem = (string & {}) | UnknownItem

interface BaseMusicItem {
	id: number
	name: string
}

export interface ParsedTrackData {
	name: string
	album: StringOrUnknownItem
	artists: StringOrUnknownItem[]
	year: StringOrUnknownItem
	duration: number
	genre: string[]
	trackNo?: number
	trackOf?: number
	image?: {
		optimized: boolean
		small: Blob
		full: Blob
	}
	primaryColor?: number
}

export interface UnknownTrack extends ParsedTrackData {
	uuid: string
	file: FileEntity
	scannedAt: number
	fileName: string
	directory: number
}

export interface Track extends BaseMusicItem, UnknownTrack {}

export interface Album extends BaseMusicItem {
	uuid: string
	artists: string[]
	year?: string
	image?: Blob
}

export interface Artist extends BaseMusicItem {
	uuid: string
}

export interface Playlist extends BaseMusicItem {
	uuid: string
	description: string
	createdAt: number
}

export interface PlaylistEntry {
	id: number
	playlistId: number
	trackId: number
	addedAt: number
}

export interface Directory {
	id: number
	handle: FileSystemDirectoryHandle
}
