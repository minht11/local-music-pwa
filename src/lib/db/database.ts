import type { DBSchema, IDBPDatabase, IDBPObjectStore, IndexNames, StoreNames } from 'idb'
import { openDB } from 'idb'
import type {
	Album,
	Artist,
	Directory,
	Playlist,
	PlaylistEntry,
	Track,
} from '$lib/library/types.ts'

export interface AppDB extends DBSchema {
	tracks: {
		key: number
		value: Track
		indexes: Pick<
			Track,
			| 'uuid'
			| 'name'
			| 'album'
			| 'year'
			| 'duration'
			| 'artists'
			| 'directory'
			| 'fileName'
			| 'scannedAt'
		> & {
			path: [directoryId: number, fileName: string]
		}
		meta: {
			notAllowedOperations: undefined
		}
	}
	albums: {
		key: number
		value: Album
		indexes: Pick<Album, 'uuid' | 'name' | 'artists' | 'year'>
		meta: {
			notAllowedOperations: undefined
		}
	}
	artists: {
		key: number
		value: Artist
		indexes: Pick<Artist, 'uuid' | 'name'>
		meta: {
			notAllowedOperations: undefined
		}
	}
	playlists: {
		key: number
		value: Playlist
		indexes: Pick<Playlist, 'uuid' | 'name' | 'createdAt'>
		meta: {
			notAllowedOperations: undefined
		}
	}
	playlistEntries: {
		key: number
		value: PlaylistEntry
		indexes: Pick<PlaylistEntry, 'playlistId' | 'trackId' | 'addedAt'> & {
			playlistTrack: [playlistId: number, trackId: number]
		}
		meta: {
			notAllowedOperations: 'update'
		}
	}
	directories: {
		key: number
		value: Directory
		indexes: Pick<Directory, 'id'>
		meta: {
			notAllowedOperations: undefined
		}
	}
}

export type AppStoreNames = StoreNames<AppDB>
export type AppIndexNames<Store extends AppStoreNames> = IndexNames<AppDB, Store>

const createIndexes = <Name extends AppStoreNames>(
	store: IDBPObjectStore<AppDB, ArrayLike<AppStoreNames>, Name, 'versionchange'>,
	indexes: readonly AppIndexNames<Name>[],
	options: IDBIndexParameters = {},
) => {
	for (const name of indexes) {
		store.createIndex(name, name, options)
	}
}

const createStore = <DBTypes extends DBSchema | unknown, Name extends StoreNames<DBTypes>>(
	db: IDBPDatabase<DBTypes>,
	storeName: Name,
) =>
	db.createObjectStore(storeName, {
		keyPath: 'id',
		autoIncrement: true,
	})

const openAppDatabase = () =>
	openDB<AppDB>('snae-app-data', 1, {
		upgrade(e) {
			const { objectStoreNames } = e

			if (!objectStoreNames.contains('tracks')) {
				const store = createStore(e, 'tracks')

				createIndexes(store, ['uuid'], { unique: true })
				createIndexes(
					store,
					['name', 'album', 'year', 'duration', 'scannedAt', 'directory'],
					{
						unique: false,
					},
				)

				store.createIndex('path', ['directory', 'fileName'], {
					// We keep flat folder structure in the database
					// but in actual FS multiple files with same name
					// can exist in different directories
					unique: false,
				})

				store.createIndex('artists', 'artists', {
					unique: false,
					multiEntry: true,
				})
			}

			if (!objectStoreNames.contains('albums')) {
				const store = createStore(e, 'albums')

				createIndexes(store, ['name', 'uuid'], { unique: true })
				createIndexes(store, ['year'])

				store.createIndex('artists', 'artists', {
					unique: false,
					multiEntry: true,
				})
			}

			if (!objectStoreNames.contains('artists')) {
				const store = createStore(e, 'artists')
				createIndexes(store, ['name', 'uuid'], { unique: true })
			}

			if (!objectStoreNames.contains('playlists')) {
				const store = createStore(e, 'playlists')
				createIndexes(store, ['uuid'], { unique: true })
				createIndexes(store, ['name', 'createdAt'])
			}

			if (!objectStoreNames.contains('playlistEntries')) {
				const store = e.createObjectStore('playlistEntries', {
					keyPath: 'id',
					autoIncrement: true,
				})

				createIndexes(store, ['playlistId', 'trackId', 'addedAt'])

				store.createIndex('playlistTrack', ['playlistId', 'trackId'])
			}

			if (!objectStoreNames.contains('directories')) {
				createStore(e, 'directories')
			}
		},
	})

type AppIDBDatabase = IDBPDatabase<AppDB>
let dbPromise: Promise<AppIDBDatabase> | AppIDBDatabase | null = null

export const getDatabase = (): Promise<AppIDBDatabase> | AppIDBDatabase => {
	if (dbPromise) {
		return dbPromise
	}

	dbPromise = openAppDatabase()

	dbPromise
		.then((db) => {
			db.onclose = () => {
				dbPromise = null
			}

			// Micro optimization to avoid unwrapping the promise
			dbPromise = db
		})
		.catch(() => {
			dbPromise = null
		})

	return dbPromise
}

export type DbKey<Name extends AppStoreNames> = AppDB[Name]['key']
export type DbValue<Name extends AppStoreNames> = AppDB[Name]['value']
