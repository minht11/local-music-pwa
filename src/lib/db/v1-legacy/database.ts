import { type DBSchema, type IDBPDatabase, openDB } from 'idb'

export const V1_LEGACY_DB_NAME = 'APP_DATA-1'

type V1LegacyTrackFileWrapper =
	| {
			type: 'fileRef'
			file: FileSystemFileHandle
	  }
	| {
			type: 'file'
			file: File
	  }

/**
 * This types does not contain all of the fields that exist,
 * but only those needed for migration.
 */
interface V1LegacyTrack {
	id: string
	fileWrapper: V1LegacyTrackFileWrapper
}

/**
 * This types does not contain all of the fields that exist,
 * but only those needed for migration.
 */
interface V1LegacyPlaylist {
	dateCreated: number
	id: string
	name: string
	trackIds: string[]
}

interface V1LegacyAppDB extends DBSchema {
	'APP_DATA-1': {
		key: `data-${V1LegacyDatabaseKey}`
		value: V1LegacyTrack[] | V1LegacyPlaylist[] | string[]
	}
}

type V1LegacyDatabaseKey = 'tracks' | 'playlists' | 'favorites'
type V1LegacyDatabaseDataMap = {
	tracks: Record<string, V1LegacyTrack>
	playlists: Record<string, V1LegacyPlaylist>
	favorites: string[]
}

/** Database used in V1 version of the app */
export const getV1LegacyDatabase = (onUpgraded?: (db: IDBPDatabase<V1LegacyAppDB>) => void) =>
	openDB<V1LegacyAppDB>(V1_LEGACY_DB_NAME, 1, {
		upgrade(db) {
			onUpgraded?.(db)
		},
	})

export const getV1LegacyDatabaseValue = async <K extends V1LegacyDatabaseKey>(key: K) => {
	const db = await getV1LegacyDatabase()
	const data = await db.get(V1_LEGACY_DB_NAME, `data-${key}`)

	return data as V1LegacyDatabaseDataMap[K] | undefined
}

export const removeV1LegacyDatabase = (): void => {
	indexedDB.deleteDatabase(V1_LEGACY_DB_NAME)
}

export const checkForV1LegacyDatabaseData = async (): Promise<boolean> => {
	const dbList = await indexedDB.databases()
	const dbExists = dbList.some((db) => db.name === V1_LEGACY_DB_NAME)
	if (!dbExists) {
		return false
	}

	const tracks = await getV1LegacyDatabaseValue('tracks')
	if (!tracks) {
		return false
	}

	if (Object.keys(tracks).length === 0) {
		try {
			void removeV1LegacyDatabase()
		} catch {
			// Ignore errors
		}

		return false
	}

	return true
}
