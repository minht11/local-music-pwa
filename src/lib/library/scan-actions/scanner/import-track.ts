import type { IDBPTransaction } from 'idb'
import { type AppDB, getDatabase } from '$lib/db/database.ts'
import { type DatabaseChangeDetails, dispatchDatabaseChangedEvent } from '$lib/db/events.ts'
import {
	type Album,
	type Artist,
	type Track,
	UNKNOWN_ITEM,
	type UnknownTrack,
} from '$lib/library/types.ts'

type ImportTrackTx = IDBPTransaction<
	AppDB,
	('tracks' | 'albums' | 'artists' | 'playlistEntries')[],
	'readwrite'
>

const dbImportAlbum = async (tx: ImportTrackTx, track: Track) => {
	const albumName = track.album

	const store = tx.objectStore('albums')

	const existingAlbum = await store.index('name').get(albumName)
	const updatedAlbum: Omit<Album, 'id'> = existingAlbum
		? {
				...existingAlbum,
				artists: [...new Set([...existingAlbum.artists, ...track.artists])].filter(
					(artist) => artist !== UNKNOWN_ITEM,
				),
				year: existingAlbum.year ?? track.year,
				image: existingAlbum.image ?? track.image?.full,
			}
		: {
				uuid: crypto.randomUUID(),
				name: albumName,
				artists: track.artists,
				year: track.year,
				image: track.image?.full,
			}

	const albumId = await store.put(updatedAlbum as Album)

	const change: DatabaseChangeDetails = {
		storeName: 'albums',
		key: albumId,
		operation: existingAlbum ? 'update' : 'add',
	}

	return change
}

const dbImportArtist = async (tx: ImportTrackTx, artistName: string) => {
	const store = tx.objectStore('artists')

	const existingArtistId = await store.index('name').getKey(artistName)
	if (existingArtistId) {
		return
	}

	const newArtist: Omit<Artist, 'id'> = {
		name: artistName,
		uuid: crypto.randomUUID(),
	}

	const artistId = await store.put(newArtist as Artist)

	const change: DatabaseChangeDetails = {
		storeName: 'artists',
		key: artistId,
		operation: 'add',
	}

	return change
}

const dbImportArtists = (tx: ImportTrackTx, artistNames: string[]) =>
	Promise.all(artistNames.map(async (artist) => dbImportArtist(tx, artist)))

export const dbImportTrack = async (
	metadata: UnknownTrack,
	existingTrackId: number | undefined,
): Promise<number> => {
	const db = await getDatabase()
	const tx = db.transaction(['tracks', 'albums', 'artists', 'playlistEntries'], 'readwrite')

	const trackId = await tx.objectStore('tracks').put(metadata as Track, existingTrackId)
	const track: Track = {
		...metadata,
		id: trackId,
	}

	const [albumChange, artistsChanges] = await Promise.all([
		dbImportAlbum(tx, track),
		dbImportArtists(tx, track.artists),
		tx.done,
	])

	dispatchDatabaseChangedEvent([
		{
			storeName: 'tracks',
			key: trackId,
			operation: existingTrackId === trackId ? 'update' : 'add',
		},
		albumChange,
		...artistsChanges,
	])

	return trackId
}
