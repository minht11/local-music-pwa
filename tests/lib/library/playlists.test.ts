import 'fake-indexeddb/auto'
import { beforeEach, describe, expect, it, vi } from 'vitest'
import { getDatabase } from '$lib/db/database.ts'
import {
	createPlaylist,
	dbAddTracksToPlaylistsWithTx,
	dbBatchModifyPlaylistsSelection,
	dbCreatePlaylist,
	dbRemovePlaylist,
	dbRemoveTracksFromPlaylistsWithTx,
	getPlaylistEntriesDatabaseStore,
	removeTrackEntryFromPlaylist,
	toggleFavoriteTrack,
	type UpdatePlaylistOptions,
	updatePlaylist,
} from '$lib/library/playlists-actions.ts'
import { dbImportTrack } from '$lib/library/scan-actions/scanner/import-track.ts'
import { FAVORITE_PLAYLIST_ID, type UnknownTrack } from '$lib/library/types.ts'
import { clearDatabaseStores, dbGetAllAndExpectLength, expectToBeDefined } from '../../shared.ts'

vi.mock('$lib/components/snackbar/snackbar', () => ({
	snackbar: Object.assign(vi.fn(), {
		unexpectedError: vi.fn(),
	}),
}))

let uuidCounter = 0

vi.stubGlobal('crypto', {
	randomUUID: vi.fn(() => {
		uuidCounter += 1
		return `test-uuid-${uuidCounter}`
	}),
})

vi.stubGlobal('Date', {
	now: vi.fn(() => 1234567890),
})

let trackCounter = 0

const dbImportTestTrack = async (overrides: Partial<UnknownTrack> = {}): Promise<number> => {
	trackCounter += 1
	const trackData: UnknownTrack = {
		uuid: `test-track-uuid-${trackCounter}`,
		name: `Test Track ${trackCounter}`,
		album: 'Test Album',
		artists: ['Test Artist'],
		year: '2023',
		duration: 180,
		genre: ['Rock'],
		file: new File(['test'], 'test.mp3', { type: 'audio/mp3' }) as UnknownTrack['file'],
		scannedAt: Date.now(),
		fileName: `test-${trackCounter}.mp3`,
		directory: 1,
		...overrides,
	}

	return await dbImportTrack(trackData, undefined)
}

describe('playlists', () => {
	beforeEach(async () => {
		await clearDatabaseStores()
		vi.clearAllMocks()
		trackCounter = 0
		uuidCounter = 0
	})

	describe('playlist creation', () => {
		it('creates new playlist with all fields', async () => {
			await dbCreatePlaylist('Test Playlist', 'My description')

			const db = await getDatabase()
			const [playlist] = await db.getAll('playlists')

			expect(playlist?.name).toBe('Test Playlist')
			expect(playlist?.description).toBe('My description')
			expect(playlist?.uuid).toBe('test-uuid-1')
			expect(playlist?.createdAt).toBe(1234567890)
		})
	})

	describe('UI wrapper functions', () => {
		it('creates playlist via UI wrapper', async () => {
			await createPlaylist('UI Playlist', 'Created via UI')

			await dbGetAllAndExpectLength('playlists', 1)

			const db = await getDatabase()
			const [playlist] = await db.getAll('playlists')

			expectToBeDefined(playlist)
			expect(playlist.name).toBe('UI Playlist')
			expect(playlist.description).toBe('Created via UI')
		})

		it('removes track entry from playlist via UI action', async () => {
			const trackId = await dbImportTestTrack()
			const playlistId = await dbCreatePlaylist('Test Playlist', '')

			const store = await getPlaylistEntriesDatabaseStore()
			await dbAddTracksToPlaylistsWithTx(store, {
				playlistIds: [playlistId],
				trackIds: [trackId],
			})

			const db = await getDatabase()
			const [entry] = await db.getAll('playlistEntries')
			expectToBeDefined(entry)

			await removeTrackEntryFromPlaylist(entry.id)

			await dbGetAllAndExpectLength('playlistEntries', 0)
		})
	})

	describe('playlist updates', () => {
		it('updates playlist name and description', async () => {
			const playlistId = await dbCreatePlaylist('Original Name', 'Original description')

			const updateOptions: UpdatePlaylistOptions = {
				id: playlistId,
				name: 'Updated Name',
				description: 'Updated description',
			}

			const result = await updatePlaylist(updateOptions)

			expect(result).toBe(true)

			const db = await getDatabase()
			const [playlist] = await db.getAll('playlists')

			expectToBeDefined(playlist)
			expect(playlist.name).toBe('Updated Name')
			expect(playlist.description).toBe('Updated description')
			expect(playlist.uuid).toBe('test-uuid-1')
			expect(playlist.createdAt).toBe(1234567890)
		})

		it('fails to update non-existent playlist', async () => {
			const updateOptions: UpdatePlaylistOptions = {
				id: 999,
				name: 'Non-existent',
				description: 'Should fail',
			}

			const result = await updatePlaylist(updateOptions)

			expect(result).toBe(false)
		})
	})

	describe('playlist removal', () => {
		it('removes playlist and associated entries', async () => {
			const trackId = await dbImportTestTrack()
			const playlistId = await dbCreatePlaylist('Test Playlist', '')

			const store = await getPlaylistEntriesDatabaseStore()
			await dbAddTracksToPlaylistsWithTx(store, {
				playlistIds: [playlistId],
				trackIds: [trackId],
			})

			await dbGetAllAndExpectLength('playlistEntries', 1)

			await dbRemovePlaylist(playlistId)

			await dbGetAllAndExpectLength('playlists', 0)
			await dbGetAllAndExpectLength('playlistEntries', 0)
		})

		it('removes only entries for specific playlist', async () => {
			const trackId = await dbImportTestTrack()
			const playlist1Id = await dbCreatePlaylist('Playlist 1', '')
			const playlist2Id = await dbCreatePlaylist('Playlist 2', '')

			const store = await getPlaylistEntriesDatabaseStore()
			await dbAddTracksToPlaylistsWithTx(store, {
				playlistIds: [playlist1Id, playlist2Id],
				trackIds: [trackId],
			})

			await dbGetAllAndExpectLength('playlistEntries', 2)

			await dbRemovePlaylist(playlist1Id)

			await dbGetAllAndExpectLength('playlists', 1)
			const [remainingEntry] = await dbGetAllAndExpectLength('playlistEntries', 1)
			expectToBeDefined(remainingEntry)
			expect(remainingEntry.playlistId).toBe(playlist2Id)
		})
	})

	describe('playlist entries management', () => {
		it('adds tracks to multiple playlists', async () => {
			const track1Id = await dbImportTestTrack({ name: 'Track 1' })
			const track2Id = await dbImportTestTrack({ name: 'Track 2' })
			const playlist1Id = await dbCreatePlaylist('Playlist 1', '')
			const playlist2Id = await dbCreatePlaylist('Playlist 2', '')

			const store = await getPlaylistEntriesDatabaseStore()
			const changes = await dbAddTracksToPlaylistsWithTx(store, {
				playlistIds: [playlist1Id, playlist2Id],
				trackIds: [track1Id, track2Id],
			})

			expect(changes).toHaveLength(4)
			await dbGetAllAndExpectLength('playlistEntries', 4)

			const db = await getDatabase()
			const entries = await db.getAll('playlistEntries')
			const playlist1Entries = entries.filter((e) => e.playlistId === playlist1Id)
			const playlist2Entries = entries.filter((e) => e.playlistId === playlist2Id)

			expect(playlist1Entries).toHaveLength(2)
			expect(playlist2Entries).toHaveLength(2)
		})

		it('removes tracks from specific playlists', async () => {
			const track1Id = await dbImportTestTrack({ name: 'Track 1' })
			const track2Id = await dbImportTestTrack({ name: 'Track 2' })
			const playlist1Id = await dbCreatePlaylist('Playlist 1', '')
			const playlist2Id = await dbCreatePlaylist('Playlist 2', '')

			let store = await getPlaylistEntriesDatabaseStore()
			await dbAddTracksToPlaylistsWithTx(store, {
				playlistIds: [playlist1Id, playlist2Id],
				trackIds: [track1Id, track2Id],
			})

			await dbGetAllAndExpectLength('playlistEntries', 4)

			store = await getPlaylistEntriesDatabaseStore()
			const changes = await dbRemoveTracksFromPlaylistsWithTx(store, {
				playlistIds: [playlist1Id],
				trackIds: [track1Id],
			})

			expect(changes).toHaveLength(1)
			await dbGetAllAndExpectLength('playlistEntries', 3)

			const db = await getDatabase()
			const entries = await db.getAll('playlistEntries')
			const hasTrack1InPlaylist1 = entries.some(
				(e) => e.playlistId === playlist1Id && e.trackId === track1Id,
			)
			expect(hasTrack1InPlaylist1).toBe(false)
		})

		it('batch modifies playlist selections', async () => {
			const track1Id = await dbImportTestTrack({ name: 'Track 1' })
			const track2Id = await dbImportTestTrack({ name: 'Track 2' })
			const playlist1Id = await dbCreatePlaylist('Playlist 1', '')
			const playlist2Id = await dbCreatePlaylist('Playlist 2', '')
			const playlist3Id = await dbCreatePlaylist('Playlist 3', '')

			const store = await getPlaylistEntriesDatabaseStore()
			await dbAddTracksToPlaylistsWithTx(store, {
				playlistIds: [playlist1Id],
				trackIds: [track1Id, track2Id],
			})

			await dbGetAllAndExpectLength('playlistEntries', 2)

			const result = await dbBatchModifyPlaylistsSelection({
				trackIds: [track1Id, track2Id],
				playlistsIdsAddTo: [playlist2Id, playlist3Id],
				playlistsIdsRemoveFrom: [playlist1Id],
			})

			expect(result).toBe(true)
			await dbGetAllAndExpectLength('playlistEntries', 4)

			const db = await getDatabase()
			const entries = await db.getAll('playlistEntries')
			const playlist1Entries = entries.filter((e) => e.playlistId === playlist1Id)
			const playlist2Entries = entries.filter((e) => e.playlistId === playlist2Id)
			const playlist3Entries = entries.filter((e) => e.playlistId === playlist3Id)

			expect(playlist1Entries).toHaveLength(0)
			expect(playlist2Entries).toHaveLength(2)
			expect(playlist3Entries).toHaveLength(2)
		})

		it('returns false when no changes are made', async () => {
			const result = await dbBatchModifyPlaylistsSelection({
				trackIds: [],
				playlistsIdsAddTo: [],
				playlistsIdsRemoveFrom: [],
			})

			expect(result).toBe(false)
		})
	})

	describe('favorites functionality', () => {
		it('adds track to favorites', async () => {
			const trackId = await dbImportTestTrack()

			await toggleFavoriteTrack(false, trackId)

			const db = await getDatabase()
			const [entry] = await db.getAll('playlistEntries')

			expectToBeDefined(entry)
			expect(entry.playlistId).toBe(FAVORITE_PLAYLIST_ID)
			expect(entry.trackId).toBe(trackId)
			expect(entry.addedAt).toBe(1234567890)
		})

		it('removes track from favorites', async () => {
			const trackId = await dbImportTestTrack()

			await toggleFavoriteTrack(false, trackId)
			await dbGetAllAndExpectLength('playlistEntries', 1)

			await toggleFavoriteTrack(true, trackId)
			await dbGetAllAndExpectLength('playlistEntries', 0)
		})
	})

	describe('playlist entries data structure', () => {
		it('creates entries with correct structure', async () => {
			const trackId = await dbImportTestTrack()
			const playlistId = await dbCreatePlaylist('Test Playlist', '')

			const store = await getPlaylistEntriesDatabaseStore()
			await dbAddTracksToPlaylistsWithTx(store, {
				playlistIds: [playlistId],
				trackIds: [trackId],
			})

			const db = await getDatabase()
			const [entry] = await db.getAll('playlistEntries')

			expect(entry).toEqual({
				id: expect.any(Number),
				playlistId,
				trackId,
				addedAt: 1234567890,
			})
		})

		it('maintains chronological order of entries', async () => {
			const track1Id = await dbImportTestTrack({ name: 'Track 1' })
			const track2Id = await dbImportTestTrack({ name: 'Track 2' })
			const playlistId = await dbCreatePlaylist('Test Playlist', '')

			const store = await getPlaylistEntriesDatabaseStore()

			vi.mocked(Date.now).mockReturnValueOnce(1000)
			await dbAddTracksToPlaylistsWithTx(store, {
				playlistIds: [playlistId],
				trackIds: [track1Id],
			})

			vi.mocked(Date.now).mockReturnValueOnce(2000)
			await dbAddTracksToPlaylistsWithTx(store, {
				playlistIds: [playlistId],
				trackIds: [track2Id],
			})

			const db = await getDatabase()
			const entries = await db.getAll('playlistEntries')
			entries.sort((a, b) => a.addedAt - b.addedAt)

			expect(entries).toHaveLength(2)
			expect(entries[0]?.trackId).toBe(track1Id)
			expect(entries[0]?.addedAt).toBe(1000)
			expect(entries[1]?.trackId).toBe(track2Id)
			expect(entries[1]?.addedAt).toBe(2000)
		})
	})
})
