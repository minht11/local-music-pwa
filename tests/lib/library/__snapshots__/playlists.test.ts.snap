// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`playlists > create new playlist 1`] = `
{
  "createdAt": 1234567890,
  "description": "",
  "id": 1,
  "name": "Test Playlist",
  "uuid": "test-uuid",
}
`;
